content/posts/looker-pdt-to-dbt-part-one.md

```diff
@@ -0,0 +1,184 @@
+	a.id,
+	a.name
+from salesforce.sf_account a
+left join salesforce.sf_opportunity o
+	on o.accountid = a.id
+
+{{< /highlight >}}
+
+The ID column was supposed to be your unique primary key and now it isn't. There is currently no automated mechanism in Looker to test the output of that query. No way of knowing that a record for ID `176` is now there 5 times, one for each opportunity generated for that account.
+
+In dbt, you can define constraints on the model, giving you the ability to identify instances like these in your development workflow.
+
+{{< highlight yaml >}}
+
+- name: sf_account
+  columns:
+    - name: id
+      tests:
+        - not_null
+        - unique
+
+{{< /highlight >}}
+
+When you run `dbt test`, you'd realise your mistake and fix it before your Head of Sales starts shouting at you because all the numbers changed. Though testing of that sort hasn't always been common-place in analytics, it's incredibly important so that everyone can have confidence in the reporting they consume.
+
+## PDT table names
+
+David from your data science team wants to build a fancy model, predicting how many flying mattresses you'll sell next week. To do so, he wants to use the `customer_status` dimension defined in your `customers` view in Looker. He doesn't want to have re-define it in his work. He understandably thinks you should all be using the same definition. 
+
+He jumps into his database explorer to find the name of the PDT in the `looker_scratch` schema, the location all the PDTs are built. He finds it. The table is called `lr$zzzxu3bfa1lkc2x5ne75_customers`. He goes off and builds his model. It has an AUC score of 0.96. He's incredibly pleased. He shares the analysis. 
+
+Your COO opens it up, tries to run it. To his (and David's) dismay, the following error occurs: `ERROR: relation "looker_scratch.lr$zzzxu3bfa1lkc2x5ne75_customers" does not exist`. The table doesn't exist anymore. 
+
+Every time you make a change to the SQL in a PDT, Looker will rename the underlying table in your database. As is the case here, Looker is often not the only platform or location where data analysis will be produced in your business. Often you'll have a data science function. Often you'll have people who need to write ad-hoc sql queries for specific pieces of analysis. As David initially asserted, all those functions should be building off the same models and the same definitions, building off the same single source of truth. 
+
+The way Looker builds its PDTs isn't conducive to that principal. The table names will continue to change (and aren't memorable for analysts to use in ad-hoc querying). 
+
+dbt will build your `customers` model in a table called `customers`. That will never change unless you tell it to. It can be consumed by every user and every platform in perpetuity. David can use it for his analysis. He could even put that model into production because he can be confident that your analytics team controls if that table exists or doesn't.
```

Comment:
I feel there needs to be a mention of a schema here, not sure if using `target` is too dbt-jargon-y
```suggestion
 dbt will build your `customers` model in a table called `customers` in your target schema. That will never change unless you tell it to. It can be consumed by every user and every platform in perpetuity. David can use it for his analysis. He could even put that model into production because he can be confident that your analytics team controls if that table exists or doesn't.
```

Reply:
Given both push to their own schemas, I kinda assumed there is parity of understanding and it didn't need to be mentioned?

Reply:
Actually, had added and amended slightly.